kafkatail/kafkatail.go

```diff
@@ -2,63 +2,90 @@ package kafkatail
 	Topic     string `long:"topic" description:"kafka topic" default:"my_topic"`
 	Partition int32  `long:"partition" description:"partition to read from"`
+	Offset    int64  `long:"offset" description:"offset to start from" default:"sarama.OffsetNewest"`
 }
 
 // GetChans returns a list of channels but it only ever has one entry - the
 // partition on which we're listening.
 // TODO listen on multiple channels to multiple partitions
-func GetChans(ctx context.Context, options Options) ([]chan string, error) {
-	linesChans := make([]chan string, 1, 1)
+func GetChans(ctx context.Context, options Options) ([]chan string, chan error) {
+	linesCh := make([]chan string, 1, 1)
 	lines := make(chan string, 1)
-	linesChans[0] = lines
+	linesCh[0] = lines
+
+	config := sarama.NewConfig()
+	config.Consumer.Return.Errors = true
+
+	brokers := []string{options.Server + ":" + options.Port}
 
-	serverString := options.Server + ":" + options.Port
-	// TODO use a reasonable kafka *Config instead of nil for the new consumer
-	consumer, err := sarama.NewConsumer([]string{serverString}, nil)
+	consumer, err := sarama.NewConsumer(brokers, config)
 	if err != nil {
 		panic(err)
 	}
 
-	partitionConsumer, err := consumer.ConsumePartition(
-		options.Topic, options.Partition, sarama.OffsetNewest)
+	partitionConsumer, err := consumer.ConsumePartition(options.Topic, options.Partition, options.Offset)
 	if err != nil {
 		panic(err)
 	}
 
+	errCh := make(chan error, 1)
+
 	go func() {
-		log.Printf("consumer started")
+		log.Printf("Consumer started for topic %s\n", options.Topic)
+
+		defer func() {
+			log.Printf("Terminating consumer for topic %s\n", options.Topic)
+			close(lines)
+
+			if err := partitionConsumer.Close(); err != nil {
+				log.Fatalln(err)
+			}
+
+			if err := consumer.Close(); err != nil {
+				log.Fatalln(err)
+			}
+		}()
+
+		// TODO: Make these configurable
+		expBackoff := backoff.NewExponentialBackOff()
+		expBackoff.MaxInterval = 30 * time.Second
+		expBackoff.MaxElapsedTime = 5 * time.Minute
+
+		cb := circuit.NewConsecutiveBreaker(10)
+		cb.BackOff = expBackoff
+
 		for {
-			select {
-			case msg := <-partitionConsumer.Messages():
-				if msg != nil {
+			if cb.Ready() {
+				select {
+				case err := <-partitionConsumer.Errors():
+					log.Printf("Error at offset %d in topic %s: %v\n",
+						partitionConsumer.HighWaterMarkOffset()-1, options.Topic, err)
+					cb.Fail()
+
+					if cb.Tripped() && expBackoff.NextBackOff() == backoff.Stop {
+						errCh <- err
+					}
+				case msg := <-partitionConsumer.Messages():
+					cb.Success()
 					lines <- string(msg.Value)
-				} else {
-					time.Sleep(1000 * time.Millisecond)
-				}
-			case <-ctx.Done():
-				// listen for the context's Done channel to clean up and exit
-				close(lines)
-				if err := partitionConsumer.Close(); err != nil {
-					log.Fatalln(err)
-				}
-				if err := consumer.Close(); err != nil {
-					log.Fatalln(err)
+				case <-ctx.Done():
+					return
```

Comment:
It's not clear to me that this loop terminates when messages or errors is closed; it looks like receiving from `ctx.Done` is the only way this routine stops. Are we expecting whoever is listening on `errCh` to kill us?

Reply:
That’s not what I see. https://github.com/cargurus/honeykaf/blob/6614d75a03547f2746b124e1f93bec983b754ae0/hk.go#L81